### DISCUSSION TITLE
성공/실패 응답 데이터 구조를 통일하는게 좋을까요?

### DISCUSSION THREAD

**Discussion by Starlight258** [Attica-org/athens-backend#41] [General]
PR : https://github.com/Attica-org/athens-backend/pull/40

안녕하세요 아테네 팀 여러분 :) 🙋
최근에 성공, 실패 응답을 통일하는 것이 좋을지에 대해 이야기가 나왔습니다. 
이에 대해 두 가지 의견이 있었는데요, 정리해 보겠습니다.

## 응답 구조를 통일해야 한다는 의견:

성공, 실패 응답 모두 일관된 구조를 가져야 클라이언트에서 응답 처리하기 편할 것 같아요.
success 필드는 성공, 실패 응답 구조를 함께 표현하기 위한 필드입니다.
응답 구조를 성공, 실패 시 동일하게 가져갈 것인지는 프론트엔드 팀과도 함께 상의해 보는 게 좋겠습니다.

## 응답 구조를 통일하지 않아도 된다는 의견:

에러 응답이 공통 응답 스펙에 맞춰야 할 필요가 있을까요?
예외 처리 시 success 필드는 항상 false를 반환할 텐데, 이게 정말 의미 있는 필드인지 고민됩니다.
만약 응답이 달라져도 괜찮다면, success 필드는 제안해 주신 것처럼 필요가 없을 것 같네요.

### 추가로 제안된 사항들은 다음과 같습니다:

에러 코드를 응답 구조에 포함시키는 것이 좋지 않을까요?
에러 메시지는 영어로 통일하는 게 어떨까요?
status와 statusCode가 중복되는 정보를 담고 있어서 상태 코드만 포함해도 좋을 것 같아요.

이에 대한 여러분의 의견을 듣고 싶습니다. 
성공, 실패 응답 구조를 통일할 것인지, 그리고 제안된 사항들에 대해서도 자유롭게 의견을 나누어 주시면 감사하겠습니다.

### 응답 구조 통일할 경우
<img width="450" alt="image" src="https://github.com/Attica-org/athens-backend/assets/78211281/23058dfa-b8bd-4825-b716-e8836b0c1657">

### 성공, 실패 응답 다르게 가져갈 경우
![image](https://github.com/Attica-org/athens-backend/assets/78211281/d4ac7cb9-e544-44e1-8395-af36b3251ee1)

**Comment by H-sooyeon** [comment]
저는 통일하는게 더 일관적이라고 생각합니다.
나중에 유지보수하면서 더 고민해보고 구조를 다르게 가야할 것 같다 싶을 때 바꾸는게 좋을 것 같습니다.

**Reply by Starlight258** [reply]
응답 구조를 일관적으로 가져갔을 때 프론트측에서 처리가 더 쉽나요?
코드가 더 깔끔해진다던지, 유지보수가 편하다던지.. 등등의 장점이 있는지 궁금합니다.

**Comment by big-cir** [comment]
저는 에러 응답이 공통 응답 스펙에 맞춰야야 할 필요가 있을까? 라는 고민을 했습니다. 
예를들어 예외에서 핸들링되는 응답의 success 필드는 항상 false를 반환 할텐데, 이게 정말 의미있는 필드인가?

결국 응답 스펙은 클라이언트와 서버와의 소통이기 때문에 적절히 조율하면 될 것 같습니다.

**Reply by Starlight258** [reply]
카카오 같은 경우에는 실패시 에러 메세지와 에러 코드만 전달하네요
![image](https://github.com/Attica-org/athens-backend/assets/78211281/b275fe24-b4fb-4edc-9fdf-9412bd4314ee)
https://developers.kakao.com/docs/latest/ko/kakaologin/rest-api

**Reply by Starlight258** [reply]
네이버도 제공하는 api에서는 에러 메세지를 에러 메세지와 에러 코드만을 전달하네요
상태 코드는 따로 전달하지 않는 것을 보니, HTTP 상의 상태 코드와 에러 메세지로 에러를 판단하게 하네요
![image](https://github.com/Attica-org/athens-backend/assets/78211281/ecabfbb5-b2e4-4051-9581-4c440292cec7)
[NAVER API](https://developers.naver.com/docs/common/openapiguide/errorcode.md#%EC%98%A4%EB%A5%98-%EB%A9%94%EC%8B%9C%EC%A7%80-%ED%98%95%EC%8B%9D)

**Reply by big-cir** [reply]
아래 @yybmion 님 말씀처럼 예상치 못한 부분에서 확장성이 떨어질 수 있으나,
저는 클라이언트의 개발 용이성도 중요한 부분인 것 같습니다.

**Comment by yybmion** [comment]
저는 특정 API 호출은 다른 호출보다 더 많은 정보를 요구할 경우가 있다고 생각하는데 그런 경우에는 표준화된 응답 형식이 제한적일 수 있다고 생각합니다.  쉽게 말해 응답 구조에 유연성을 제공하면 좋다고 생각합니다.

**Reply by Starlight258** [reply]
응답이 JSON 형식이므로 일관된 응답 구조 안에서도 충분히 유연하게 확장가능하지 않을까요?

**Comment by H-sooyeon** [comment]
api마다 응답 구조를 다르게 가면 프론트에서 응답을 처리하기 불편합니다.
나중에 유지보수할 때도 보는게 쉽지 않아요.

**Reply by Starlight258** [reply]
아 그렇군요 ! 코드가 더 단순해지는 장점이 있네요

**Comment by H-sooyeon** [comment]
네이버페이는 아래처럼 기본 응답 필드를 정해두고 있습니다. <br />
<img width="989" alt="image" src="https://github.com/Attica-org/athens-backend/assets/56586470/26243dd1-2ab6-47ed-9cb0-aab0219a4c8c">

<img width="875" alt="image" src="https://github.com/Attica-org/athens-backend/assets/56586470/55b1c48e-116b-411f-b813-b3baef2c1cde">

**Reply by Starlight258** [reply]
오 그렇군요, api 구현 방식이 회사별로도 나뉘네요!

**Comment by yybmion** [comment]
정해진 답은 당연히 없으니 클라이언트단을 고려하여 저희가 생각한 `응답에 필요한 필드`를 조합해서 통일시키는게 좋을 것 같습니다.

**Comment by Starlight258** [comment]
여러분 의견 다 나누셨을까용 
 디코로 모여주시길 바랍니더
@H-sooyeon @big-cir @yybmion

**Comment by Starlight258** [comment]
결과 : 응답 구조를 통일한다.

### 예: API 호출 성공
```java
{
    "success": true, // API 호출 성공
    "data": { // Payload: 성공 시 요청한 데이터

     }
}
```

### API 호출 실패
```java
{
    "success": false, // API 호출 실패
    "data": { // Payload: 실패 시, 실패에 관한 상세 정보
       "errorCode": 024,
        "errorMessage": "error message" // 메시지
    }
} 
```
